### ISSUE TITLE
kcov returns incorrect exit code on abort signal

### ISSUE THREAD

**Issue by matdoering** [SimonKagstrom/kcov#314]
When kcov is called on an executable that terminates with signal 6, the error is not propagated to the return code of kcov. Instead, kcov simply returns an RC of 0.

Consider the following example (I wasn't able to construct a minimal example, so I'm just showing the debug 15 output from running kcov on a google test suite that produces the error):

> [ FATAL ] /tmp/googletest-release-1.8.0/googletest/src/gtest-port.cc:1020:: Only one stdout capturer can exist at a time.
> PT stopped PID 130 0x0000067f
> PT signal 6 at 0x7fffee28b336 for 130
> PT continuing 130 with signal 6
> PT stopped PID 157 0x00000000
> PT exit 0 at 0xffffffffffffffff for 157
> PT continuing 157 with signal 0
> PT error for 157: -1
> PT stopped PID 130 0x00000086
> PT terminating signal 6 at 0x7fffee28b336 for 130
> PT continuing 130 with signal 0
> PT error for 130: -1
> Returning error

After that 

> echo $? 

returns a value of 0. This is interesting because kcov seems to recognize the error and even states "Returning error" in the debug output.

Why is fixing this a high priority? When kcov is integrated into a build pipeline running using continuous integration, it should not miss failing tests. In this case, kcov misses the failing test but without kcov, the failure is detected because the correct exit code is returned.

I just took a quick dive into the code and I think what may cause the issue is this part:

https://github.com/SimonKagstrom/kcov/blob/afa0acea5ba9c7f14cd92a687d212ece28340a02/src/engines/ptrace.cc##L295-L296

From what I see, _ev.data_ has the correct error code set from _waitEvent()_. However, the next line in the code checks for  _ev_signal_ and only in this case uses ev.data and otherwise 0. I think that _ev.data_ should also be used for other signals than ev_signal (e.g. including sigabort). What do you think?

**Comment by SimonKagstrom** [comment]
Thanks for the detailed and very good analysis!

Actually, I think the problem might be caused by the combination of your find and here:

https://github.com/SimonKagstrom/kcov/blob/master/src/collector.cc#L118

which only does an exit if the event type is ev_signal_exit. That in turn is perhaps caused by

https://github.com/SimonKagstrom/kcov/blob/master/src/engines/ptrace.cc#L249

i.e., if there are more threads/forked processes left it doesn't set ev_signal_exit. It should probably do that on e.g., the abort signal.

**Comment by dsteeley** [comment]
We've had a similar issue when kcov is called on an executable that terminates with signal 11 (SEGFAULT). Again the return code isn't propagated and kcov returns a value of 0. This is mainly an issue as kcov is used in a build pipeline that should fail but doesn't. 

I've looked at the code mentioned previously in this thread and have identified a potential issue and solution. 
In [https://github.com/SimonKagstrom/kcov/blob/48a300734904e48d1300e54e4d45ddaf0a823faf/src/collector.cc#L120](https://github.com/SimonKagstrom/kcov/blob/master/src/collector.cc#L120)
The exit code `m_exitCode` isn't set when crashing out with an `ev_signal_exit` so we don't pass the raised signal up. 
By adding the same as the exit case below, `m_exitCode = ev.data;` when a SEGFAULT is hit in the executed code kcov will return it. 

I've also been unable to construct an example but here is the debug=15 output
```
PT continuing 11759 with signal 11
PT stopped PID 11759 0x00000b7f
PT signal 11 at 0x7ffff66624e0 for 11759
PT continuing 11759 with signal 11
PT stopped PID 11762 0x0000000b
PT terminating signal 11 at 0x7ffff66624e0 for 11762
PT continuing 11762 with signal 11
PT error for 11762: -1
```
... threads all get terminated 
```
PT stopped PID 11755 0x0000000b
PT terminating signal 11 at 0x7ffff66624e0 for 11755
PT continuing 11755 with signal 0
PT error for 11755: -1
Returning error
```
Where `echo $?` doesn't indicate an error.

I've found that for each thread being terminated ev.type is set to ev_signal except the final thread which does get ev.type set to ev_signal_exit.

Is setting `m_exitCode` and in the `ev_signal_exit` case of  `onEvent` an appropriate solution?
